Register Navigation Link block variations based on menu item types added in customizer filters

Aljullu opened the pull request:
## Description
This PR adds support for the `customize_nav_menu_available_items` and `customize_nav_menu_available_item_types` filters in the navigation block and navigation screen.

Fixes #31316.

In order to do so:
- A new Navigation Link block variation is registered for each menu custom item type returned from `customize_nav_menu_available_item_types`.
- A new endpoint has been created so items added in `customize_nav_menu_available_items` can be queried by the LinkControl component.
- I fixed some typos and added some missing documentation here and there.

Note: I didn't include any tests yet, before I would like to validate if the approach taken in this PR makes sense.

## Testing steps

You can either test with the code snippet example I posted or with real plugins (see section below).

### With an example

Paste this PHP code somewhere, simulating it's a plugin adding some nav items:

```PHP
function add_nav_menu_item_types( $item_types ) {
	$item_types[] = array(
		'title'      => 'Gutenberg Links',
		'type_label' => 'Gutenberg Link',
		'type'       => 'gutenberg_nav',
		'object'     => 'gutenberg_link',
	);

	return $item_types;
}

function add_nav_menu_items( $items = array(), $type = '', $object = '', $page = 0 ) {
	if ( 'gutenberg_link' !== $object ) {
		return $items;
	}

	$gutenberg_items = array(
		array(
			'id'         => 'main',
			'title'      => 'Main site',
			'type_label' => 'Custom Link',
			'url'        => 'https://wordpress.org/gutenberg/',
		),
		array(
			'id'         => 'download',
			'title'      => 'Download page',
			'type_label' => 'Custom Link',
			'url'        => 'https://wordpress.org/plugins/gutenberg/',
		),
		array(
			'id'         => 'repo',
			'title'      => 'Repository',
			'type_label' => 'Custom Link',
			'url'        => 'https://github.com/WordPress/gutenberg',
		),
	);

	return array_merge( $items, $gutenberg_items );
}

add_filter( 'customize_nav_menu_available_item_types', 'add_nav_menu_item_types', 10, 1 );
add_filter( 'customize_nav_menu_available_items', 'add_nav_menu_items', 10, 4 );
```
- Create a post or page and add a Navigation block.
- Inside the Navigation block, add a Gutenberg Link block.
- Select any of the available options and verify the link is inserted correctly.

Repeat the process in the Navigation screen and widget screen with a classic theme and in the Site Editor with a block-based theme.

### With real-life plugins
<details>
<summary>WooCommerce</summary>

Currently, WooCommerce only adds the `customize_nav_menu_available_items` filter in admin screens, that makes those items not to be available in Rest API calls. I created [a branch](https://github.com/woocommerce/woocommerce/compare/fix/menu-items-in-rest) in WooCommerce's repo which would change this, so in order to test, you will need to check out that branch.

- In your `wp-content/plugins` folder clone WC (`git clone https://github.com/woocommerce/woocommerce`)
- Go inside the `woocommerce` folder and checkout the correct branch (`git checkout fix/menu-items-in-rest`)
- Build WooCommerce ([see docs](https://github.com/woocommerce/woocommerce/wiki/How-to-set-up-WooCommerce-development-environment#install-dependencies-and-generate-assets))
- Create a post or page and add a Navigation block.
- Inside the Navigation block, add a WooCommerce endpoint block.
- Select any of the available options and verify the link is inserted correctly.

Repeat the process in the Navigation screen and widget screen with a classic theme and in the Site Editor with a block-based theme.
</details>

<details>
<summary>BigCommerce</summary>

I had trouble enabling BigCommerce in a WP 5.8 site, so what I did was to install BigCommerce in a WP 5.7 and then update WP to 5.8.

Once you have BigCommerce enabled in WP 5.8 (you don't need to create an account to test this PR):

- Create a post or page and add a Navigation block.
- Inside the Navigation block, add a BigCommerce navigation link block.
- Select any of the available options and verify the link is inserted correctly.

Repeat the process in the Navigation screen and widget screen with a classic theme and in the Site Editor with a block-based theme.
</details>

### Regression testing

It would be good to test other Navigation Link variations to ensure there are no regressions in them:

- Try adding custom, post and page links to a Navigation block.
- Do the same with category and tag links.
- Make sure adding all links worked fine.

## Screenshots

_WooCommerce menu custom items in the Navigation screen:_
<img src="https://user-images.githubusercontent.com/3616980/117269542-a8f64900-ae58-11eb-8c23-3b085f9896e6.png" width="440" alt="Screenshot showing WooCommerce menu custom items"/>

_BigCommerce menu custom items in the post editor:_
<img src="https://user-images.githubusercontent.com/3616980/117269410-88c68a00-ae58-11eb-9163-0aab8e416ea5.png" width="486" alt="Screenshot showing BigCommerce menu custom items"/>

## Types of changes
New feature: register Navigation Link block variations based on the values returned by `customize_nav_menu_available_items` and `customize_nav_menu_available_item_types` filters.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

Aljullu committed:
Typos

Aljullu committed:
Add missing LinkControl prop in README

Aljullu committed:
Add Navigation Link block variations for all item types added in customize_nav_menu_available_item_types

Aljullu committed:
Create endpoint to fetch menu custom items

Aljullu committed:
Improve isCustomItemType check

getdave commented:
Thank you for raising this PR. It looks _really_ interesting and I'm pleased that we have an approach to catering for these existing filters. 

Obviously, this is quite a large PR so ultimately (not yet - let's review first) we might want to break it up into PRs for:

* REST API endpoint.
* Additional of new API to `@wordpress/core-data`.
* Modifications to `LinkControl`.
* Modifications to the Navigation block.

That should make it easier to review and (given the pace of development in Gutenberg at the moment) introduce without too many rebases along the way.

I'm going to experiment with the branch as per your instructions and start to wrap my head around what's going on and whether there are any other angles to explore. Initially, however, this seems to fit together well.

Mamaduka commented:
Hello, @Aljullu

Thanks for working on this.

I agree with @getdave that splitting this into smaller PR will make reviews easier.

One thing that I'm not certain about is if we should use `customize_nav_menu_available_*` filters. These are Customizer-specific filters and currently not support on the Menus screen.

Maybe we should introduce new navigation screen-specific filters for adding custom variations and fetching suggestions.

Aljullu commented:
Thanks for the answers @getdave and @Mamaduka!

> One thing that I'm not certain about is if we should use `customize_nav_menu_available_*` filters. These are Customizer-specific filters and currently not support on the Menus screen.
> 
> Maybe we should introduce new navigation screen-specific filters for adding custom variations and fetching suggestions.

From my understanding, there were two ways plugins were adding custom menu items (at least, that's what WooCommerce is doing):

* The `customize_nav_menu_available_items` and `customize_nav_menu_available_item_types` filters for the customizer screen.
* `add_meta_box()` inside the action `admin_head-nav-menus.php` for the menu screen.

The issue with the second approach is that it needs to add a lot of boilerplate HTML to create the meta box layout, while the first approach seems cleaner.

It's true, however, that the filter names might not be the most appropriate for the block-based navigation screen. I used those filter names to keep it backwards-compatible, but we could rename them and deprecate (but still support) the old filter names. Is that what you were proposing?

> I agree with @getdave that splitting this into smaller PR will make reviews easier.

Will do! :slightly_smiling_face:  I just want to make sure we agree on a path forward before opening new PRs for this.

Mamaduka commented:
> The issue with the second approach is that it needs to add a lot of boilerplate HTML to create the meta box layout, while the first approach seems cleaner.

Yes, the current method for adding custom menu items to the menu screen isn't ideal.

Maybe we can add `nav_menu_available_items` and `nav_menu_available_item_types` filters, that will have same signature as customizer ones.

This will make it easier for plugins to opt-in new Navigation screen support by adding another filter for the same callback.

gwwar commented:
Before we chat through implementation details, I want to better understand the problem this is solving.

So broadly, would we say we'd like for a site to define some set of links which are insertable in the navigation block? Are these links dynamic? How are they special/different from custom links and do we have any other examples besides WooCommerce?

Mamaduka commented:
> So broadly, would we say we'd like for a site to define some set of links which are insertable in the navigation block?

Yes, that's correct.

These links can be dynamic or not. Most of the time, they're inserted as "Custom Link" type. I know that BuddyPress also uses this method, and based on my quick search of the WP.org repo, there are 20 other plugins.

Search  results: https://wpdirectory.net/search/01F61B3YJ3XNDYPJS2PCYW8HEZ

Aljullu commented:
> Maybe we can add `nav_menu_available_items` and `nav_menu_available_item_types` filters, that will have same signature as customizer ones.

@Mamaduka That sounds good to me. The idea would be that the Navigation screen would only apply these new filters and extensions would need to register filters for the legacy and new hooks?

> Are these links dynamic? How are they special/different from custom links and do we have any other examples besides WooCommerce?

@gwwar Those links are not dynamic, once added they are just regular 'custom links', at least, this is the case for WooCommerce. The utility of those filters is so plugins can offer a set of links that users can choose from without the need to introduce the URL (which the user might not know at all).

getdave commented:
@Aljullu I've tried several times to get a WooCommerce build up and running but unfortunately I keep hitting memory limits. 

For ease of testing and for the benefit of other reviewers, would you be able to add some code directly to this PR (or in a comment) that utilises the functionality being added by this PR? 

It really just needs to exercise the new APIs you've created so we can review. 

I feel that's going to make the process of getting feedback here a lot easier. Let us know if you're able to do that.

Much appreciated.

Aljullu commented:
@getdave done! :+1:  I updated the PR description with a PHP snippet that can be used to test this PR, let me know if it works.

getdave commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -449,6 +465,19 @@ export default function NavigationLinkEdit( {
+					type: 'URL',
+				} ) );
+			}
+		);
```
Instead of exposing a new prop on `<LinkControl>` for `fetchSuggestions` could we bundle the fetching of custom items types into the main `fetchLinkSuggestions` API logic?

https://github.com/WordPress/gutenberg/blob/175a08fa64a17436cb341b82adaf482a26431a05/packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js#L70

We're already querying different APIs depending on various attributes such as `type` and `kind` so it feels like we could bundle this here.

Is there a reason that I've missed as to why you _didn't_ do it this way?

Aljullu replied:
> Is there a reason that I've missed as to why you _didn't_ do it this way?

Not really, I wasn't sure whether adding it to `fetchLinkSuggestions` would be a good idea. But if that sounds good to you, I'm ok doing it. And it simplifies the code a lot! :slightly_smiling_face: 

Changed in ac2adf73be05dc29b16ee0d51c312b9123db3333.

getdave commented on packages/block-library/src/navigation-link/edit.js:
```diff
@@ -242,6 +242,18 @@ export default function NavigationLinkEdit( {
 	const itemLabelPlaceholder = __( 'Add link…' );
 	const ref = useRef();
 
+	const isCustomItemType =
```
I think given that one of the built in types is `custom` it would be helpful to have a docblock here explaining more about what a "custom" link actually is.

Also perhaps consider renaming the variable to be even more explicit about [what this variable represents](https://developer.wordpress.org/reference/hooks/customize_nav_menu_available_item_types/) so as to distinguish it.

```suggestion
	const isCustomisedNavMenuItemType =
```

Finally you could consider extracting to a helper method with it's own function docblock...etc.

Aljullu replied:
With the changes in ac2adf73be05dc29b16ee0d51c312b9123db3333 this part of the code is no longer necessary.

getdave commented on lib/class-wp-rest-menu-custom-items-controller.php:
```diff
@@ -0,0 +1,78 @@
+ * Class that returns the menu items added via the
+ * `customize_nav_menu_available_items` filter.
+ */
+class WP_REST_Menu_Custom_Items_Controller extends WP_REST_Controller {
```
You'll want to tag in someone from the REST API team such as @TimothyBJacobs to take a look at this one. They have a regular Core team office hours in WP Core Slack if you need to get some traction.

spacedmonkey replied:
@getdave Can you provide context on what this endpoint does and why it is needed?

getdave replied:
Sure, although I would defer to @Aljullu who is the PR author.

The endpoint is introduced purely to retrieve any custom nav menu items that have been added by the developer via the `customize_nav_menu_available_items` filter. As you are probably aware, this filter already exists in Core.

The items are then displayed in the Nav Editor in the `<LinkControl>` search results which you can see in the screenshots attached to this PR.

@Aljullu also left a good explanation of the context behind this [in the PR description](https://github.com/WordPress/gutenberg/pull/31584#issue-632392442) and in [the accompanying Issue](https://github.com/WordPress/gutenberg/issues/31316).

As I mentioned in my review of this PR, I would hope that we will not actually need to introduce a new endpoint but rather the functionality proposed can be introduced to the main Menu Items REST API endpoint. I will however, defer to your greater experience working on these endpoints.

spacedmonkey replied:
@Aljullu @getdave I think this code should live in the menu items endpoint. The fields map pretty well. 

Maybe something along the line of. 

`/wp/v2/menu-items/custom`. 

@TimothyBJacobs WDYT?

spacedmonkey replied:
If this is the direction we are going, I think the REST API should be created in a new PR, with unit tests etc. Once it is merged, this PR can be rebased.

getdave commented on lib/class-wp-rest-menus-controller.php:
```diff
@@ -382,7 +382,7 @@ public function update_item( $request ) {
 
 		$prepared_term = $this->prepare_item_for_database( $request );
 
-		// Only update the term if we haz something to update.
```
😆

getdave commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -105,6 +105,20 @@ Whether to present suggestions when typing the URL.
 
 Whether to present initial suggestions immediately.
 
+### fetchSuggestions
```
I have concerns about exposing this as an API. Whilst it seems logical, there is actually a tonne of logic bundled up with handling input and performing the correct type of search.

We're overriding the default behaviour of the search handler in this particular scenario where we want to search for custom types. However, it really feels like searching for custom types shouldn't require a different API. As I mention elsewhere in this review, we might be able to achieve the same by modifying the existing `__experimentalFetchLinkSuggestions` API.

Aljullu replied:
This should also have been fixed with ac2adf73be05dc29b16ee0d51c312b9123db3333. :slightly_smiling_face:

getdave commented on packages/core-data/src/fetch/__experimental-fetch-menu-custom-items.js:
```diff
@@ -0,0 +1,19 @@
+import apiFetch from '@wordpress/api-fetch';
+import { addQueryArgs } from '@wordpress/url';
+
+const fetchMenuCustomItems = async ( search, { type } ) => {
```
I'm unsure about introducing yet another custom handler API here on the block editor settings. Each time we do this we have to "polyfill" it in each and every custom editor instance used in Core.  

For example

https://github.com/WordPress/gutenberg/blob/175a08fa64a17436cb341b82adaf482a26431a05/packages/edit-navigation/src/index.js#L30-L31

I'm hopeful we can avoid the need to introduce it.

Aljullu replied:
And this one too: ac2adf73be05dc29b16ee0d51c312b9123db3333.

getdave reviewed (commented):
Thank you again for this PR. It's a really great move towards better backwards compatibility for the Nav Block with the existing Menus system.

My understanding is that this PR basically allows a developer to use **_existing APIs_** in order to:

1. register additional nav menu link item _types_ which ultimately show up as new block variations of the `core/navigation-link` block. 
2. provide a set of custom _items_ to be displayed in the search _results_ when a user types into the search control when adding a link.

This is how it looks IRL:

https://user-images.githubusercontent.com/444434/121029523-9006ed80-c7a0-11eb-81e9-16f2c300a046.mov

One of the key goals of the Nav Block and Editor should be to try and provide (as close to) full backwards compatibility with the existing Menu system. For this reason, I really like the way that using the _existing_ hooks this approach doesn't require the developer to change anything in order to see their custom items in the Nav block. 

However, I'm interested to understand whether

* Whether we need a new _dedicated_ REST API endpoint to get the custom menu items. I wonder whether this could/should be achievable as part of the existing [`WP_REST_Menu_Items_Controller`](https://github.com/WordPress/gutenberg/blob/07f48a75513f62901d017e29f104e599c9b9410d/lib/class-wp-rest-menu-items-controller.php).
2. We need a new API within `@wordpress/core-data` to access the custom items. We should be able to extend the existing APIs to grab custom items.

I'd really like to get some additional input from others working on the Nav Editor and Block. 

Pinging @draganescu @Mamaduka @gwwar @talldan @tellthemachines to see if they can offer any additional opinion and insight, specifically on:

* How do they feel about the PR as it stands.
* What do they think about the points I have raised? 
* Any other ideas to improve/modify the approach?

Thanks again @Aljullu for this PR.

spacedmonkey commented on lib/class-wp-rest-menu-custom-items-controller.php:
```diff
@@ -0,0 +1,78 @@
+			foreach ( $item_types as $item_type ) {
+				if ( $item_type['type'] === $requested_type ) {
+					return rest_ensure_response(
+						apply_filters( 'customize_nav_menu_available_items', array(), $item_type['type'], $item_type['object'], 0 )
```
We need add `get_item_schema` and `prepare_item_for_response` methods.

Aljullu replied:
Done in 17d4d17a02cebd13e9aa0d998c1cec1dad061b08. I don't have a lot of experience building new endpoints, so please let me know if I didn't do it right.

Aljullu committed:
Bundle fetching of custom item types into fetchLinkSuggestions

Mamaduka commented:
Thanks for working on this, @Aljullu.

I think we're moving in the right direction.

I checked the [`load_available_items_query`](https://developer.wordpress.org/reference/classes/wp_customize_nav_menus/load_available_items_query/) method, and it does the same thing as `fetchLinkSuggestions` but for AJAX callback. So I think bundling custom item fetching into latter was a good idea.

@spacedmonkey, the new endpoint will return pre-defined custom items or custom links based on type. Users can easily select these items for a menu without the need to know complete URLs.

Do you think this is something that can be a part of the existing endpoint?

BuddyPress example:
![CleanShot 2021-06-08 at 14 39 23](https://user-images.githubusercontent.com/240569/121171648-ceb3ab00-c867-11eb-94a0-655568593270.png)

Aljullu committed:
Add get_item_schema and prepare_item_for_response methods

Aljullu commented:
Thanks for all the reviews! I think I addressed all the feedback, there are still some open questions like whether the new dedicated REST API endpoint is the best approach or if we should expand `WP_REST_Menu_Items_Controller` instead. I would like some direction on this because I don't feel entitled to make a decision on this.

> My understanding is that this PR basically allows a developer to use **_existing APIs_** in order to:
> 
> 1. register additional nav menu link item _types_ which ultimately show up as new block variations of the `core/navigation-link` block. 
> 2. provide a set of custom _items_ to be displayed in the search _results_ when a user types into the search control when adding a link.

@getdave That's correct. :ok_hand:

spacedmonkey commented on lib/class-wp-rest-menu-custom-items-controller.php:
```diff
@@ -0,0 +1,128 @@
+<?php
```
Why can't think be part of the existing menu items controller?

Also why does this have to part of the REST API at all. Can't be loaded in via localize script or similar?

getdave replied:
The results are shown in the search results within `<LinkControl>` are retrieved on demand from the REST API [using this API within Gutenberg](https://github.com/WordPress/gutenberg/blob/f3ad70f2929c4092e355cfd30afd7f1059d6bc08/packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js#L70).

It feels like the custom items should also be retrieved using the same mechanism. 

From the discussion in the PR thread it looks like there are a number of options being proposed so this may change direction.

gwwar reviewed (commented):
Thanks for looking into this @Aljullu. I'm leaving a bit of higher level feedback to 🤔 through.

Overall, I do think this is a pretty good use case to support, but I'm not sure of how this should be displayed to a user in the block editors. 

How many custom links does a plugin typically register? Is it something like a handful like 5 or a lot like 1000+? If it's closer to a handful we could maybe embed this within a script or the block variation data itself. If it's a lot, some REST endpoint makes more sense. (Whether this is new or an extended extended endpoint I'll defer to API component owners).

Another thing that could use some thought is discoverability of said custom links. By default, link suggestions only displays 3 links. How will a user know to search for items they didn't create? Maybe @jasmussen has some ideas?

https://user-images.githubusercontent.com/1270189/121207698-5977b400-c82e-11eb-925e-2b2891937aec.mp4

jasmussen commented:
This is definitely tricky from a UI point of view. As Kerry points out, the more child blocks we add, the harder it is to find the right one to insert. It's already grown to a point where the experience isn't ideal:

<img width="345" alt="Screenshot 2021-06-09 at 09 19 54" src="https://user-images.githubusercontent.com/1204802/121310865-1a1b9700-c904-11eb-8376-ff3ffb7c353e.png">

The fact that it's a two-step process: _insert block type first_, and _then choose the contents of that block type_ is likely a source of confusion for a user who inserts a navigation block and then expects the plus button to let them immediately add a link.

But that's how the block editor works — first you insert the block, and then you configure the block. Patterns is a key interface we can use to smooth this out. 

And that brings me to the next item — navigation block patterns will likely have _menu items_, search, spacers, site titles, social links preinserted. They will not pre-insert a block type that isn't part of the vanilla offering. Which means any blocks you extend with here, have to be inserted by the user as they are building their menu — they won't be able to leverage patterns.

Even now, we have both "Page Link", and "Custom Link" blocks to insert. They are technically the same, mostly — you can search for a Page in the custom Link block, or you can paste an external URL in the Page Link block:

<img width="416" alt="Screenshot 2021-06-09 at 09 19 37" src="https://user-images.githubusercontent.com/1204802/121311931-2e13c880-c905-11eb-8e52-722d58d3d17e.png">

But I'm increasingly not sure the distinction between the two is valuable, since they both do the same. And by adding additional "link type" blocks, it will only serve to make the two-step flow that much more clunky. 

Have you considered, instead of registering new link type blocks, to extend or expand the URL dialog instead? Instead of this one:

<img width="540" alt="Screenshot 2021-06-09 at 09 20 32" src="https://user-images.githubusercontent.com/1204802/121312248-7cc16280-c905-11eb-94f2-c3f259f29a61.png">

In the end, a menu item is just a text label and a link, with affordances for choosing that link. We might just want to let you pick that link interface? This is a rough sketch that needs time in the oven (perhaps something better than a dropdown to choose the type), but hopefully it illustrates the idea:

<img width="929" alt="Screenshot 2021-06-09 at 09 36 41" src="https://user-images.githubusercontent.com/1204802/121312979-3a4c5580-c906-11eb-99a6-47ee792884b8.png">

See also #24099.

Aljullu commented:
Thank you all for the feedback and ideas!

@gwwar:

> How many custom links does a plugin typically register? Is it something like a handful like 5 or a lot like 1000+?

I don't have any data, but from my tests with WooCommerce, BuddyPress and BigCommerce, it looks like they add less than a dozen links each.

> If it's closer to a handful we could maybe embed this within a script or the block variation data itself. If it's a lot, some REST endpoint makes more sense. (Whether this is new or an extended extended endpoint I'll defer to API component owners).

Do you have an example of what "within a script or the block variation data itself" would look like? That's an approach I experimented with in the beginning, but couldn't figure out how to get working. Is your proposal similar to @spacedmonkey's to use a localize script?

@jasmussen:

> Have you considered, instead of registering new link type blocks, to extend or expand the URL dialog instead?

Sure thing, that's one of the 'alternative implementations' that I considered in #31316. I didn't take that route for a couple of reasons:

* The custom menu link items are not a real post/page or any other content type. So, as a user, I might be confused on where to search those options. Following WC's example (the one that is more familiar to me): as a user I want to add a link to `Orders`, but I know there is no `Orders` page when I visit the `All Pages` admin page. Would it be clear I need to add a `Page Link` or `Custom Link` to find that option?
* This is mostly a technical one, but I didn't find any extensibility points that would allow me to hook into the `pages` API response. It should be possible to add these extensibility points, but pointing out that AFAIK they don't currently exist.
* UI-wise, I wonder if it's better to have different block variations for each plugin or to have all plugins hook into the `Page Link` or `Custom Link` variation? If a site has WooCommerce and BuddyPress installed, will it be easy to find blocks relative to each plugin? In addition to that, BuddyPress seems to separate the links between logged-in and logged-out, how could that be presented in the UI?

Notice none of these reasons are blocking and I'm completely fine exploring the route of extending the URL dialog instead of creating new Navigation Link block variations, but I wanted to expose some of the reasons that made me try the latter approach first.

jasmussen commented:
> Notice none of these reasons are blocking and I'm completely fine exploring the route of extending the URL dialog instead of creating new Navigation Link block variations, but I wanted to expose some of the reasons that made me try the latter approach first.

Thank you, that's awesome.

I'm trying to go back to the initial use case meant to be solved by this. It may have made sense for a dedicated navigation screen that is primarily intended for organizing _navigation items_. But the new navigation block isn't just links. In that context, having to pick between multiple different types of links, I'm not convinced, is the best user experience. It makes sense for the "Home Link" as that's kind of a disparate thing with little to no configuration. But even the separation between Page Link and Custom Link is spurious.

To be clear, separate blocks might still be the correct path forward for the API. But it's a path I suspect that should be paired with improvements to the quick inserter, and yes, possibly also improvements to the link dialog. I'd appreciate a wider range of thoughts here: @draganescu @mcsf if you have time, thank you.

gwwar commented:
> Do you have an example of what "within a script or the block variation data itself"

@Aljullu this depends on what flow we ultimately end up with. I'd see if we can get folks to align on a user flow in the block editor first. What's important here is that if most plugins only need a few items, both an API endpoint and some other mechanism like passing an array of links in a hook can be viable.

> In that context, having to pick between multiple different types of links, I'm not convinced, is the best user experience. It makes sense for the "Home Link" as that's kind of a disparate thing with little to no configuration. But even the separation between Page Link and Custom Link is spurious.

@jasmussen thanks for the insights here! Each custom post type may also be registered as a navigation block variant, so I wonder if we'll need to help fix this in future PRs.

Going back to this PR. One thing I'm reminded of is that since we're working with static link content, I'm actually wondering if a block pattern is a better fit. 🤔 Perhaps we could apply a pattern as an option to the initial navigation placeholder?

https://developer.wordpress.org/block-editor/reference-guides/block-api/block-patterns/

draganescu commented on packages/block-editor/src/components/link-control/README.md:
```diff
@@ -105,6 +105,13 @@ Whether to present suggestions when typing the URL.
 
 Whether to present initial suggestions immediately.
 
+### noDirectEntry
```
Where is this used?

Aljullu replied:
Ah, that's unrelated to the other changes in that PR. It's a prop that was undocumented. I thought all props had to be in the docs, but happy to remove it from here if not.

draganescu commented:
**TLDR** We should probably drop this link block variation idea altogether.

The path in this PR is basically using the same system that the default variations use, so I don't think it's wrong. 

As a user I want to add a link to "place" by adding a link block and linking it to "place". If "place" is not an URL the link block should do its best to suggest the best match, based on me entering "place" in the search box.

Based on this story, we should allow users to search for Home, Front page, Orders, Cart, Members, Profile, All forums or whatever kind of special places the system provides, and which have dynamic, generated locations. Just as the user can search in posts or pages by their title.

I don't think these variations are a great win. We did copy the way the old screen behaved, but TBH, that screen did not have a live query UI which can be extended to respond quite smartly. We now have a single entry point, a search box, the most common UX. Why do we need to feature "post link", "custom link", "orders link"? Is the win worth the problem?

We also used a variation so that we can show default suggestions that match. Say I insert a Page link I see most recent pages. But is this extra step worth it considering I have a search box which looks through all the "stuff" on my site (ideally it should search everything, including media)? Don't we just create an assumption for no reason that all "link types" (a concept which is foreign to the web) should now have a button? 

The only other problem still remaining is that all links except custom (aka manually entered) links _shoud not be static_ as they're depending on various system settings.

kevin940726 commented on packages/core-data/src/fetch/__experimental-fetch-link-suggestions.js:
```diff
@@ -153,6 +153,33 @@ const fetchLinkSuggestions = async (
+				const customItems = results.filter(
+					( result ) =>
+						search === '' ||
+						result.title.match( new RegExp( search, 'i' ) )
```
This means `search` can contain regex specific tokens such as `*` and `.`, which I don't think is what we expected. If we only want to match case-insensitivity then we can probably just use:

```js
result.title.toLowerCase() === search.toLowerCase()
```

Aljullu replied:
Good point. Fixed in 5c34f3bd8d84820e9e0a4bb9508f1bffd2d32184.

talldan commented:
Agree with the general consensus that the number of variations exposed in the inserter can be confusing.

I don't think it'd be a problem to hide most or all of them from the inserter, the type of link selected from the link interface can (and I think already does) determine the variation.

The linking interface might need some improvements to cater for the amount of items a user can search for. That could probably tie in with the possibility of bulk adding links - https://github.com/WordPress/gutenberg/issues/31667.

Aljullu committed:
Use stringToLower instead of RegExp to find endpoint matches

Aljullu commented:
Thanks for the feedback and all ideas in this PR. I applied some of the suggestions, but my understanding is that some others need some design/product decisions. I tried to gather all ideas below (please, feel free to add new ones or edit them if you feel I didn't explain them right):

- Create a Navigation link block variation. This is how it's done in this PR. Currently, data is retrieved from a newly-created endpoint, but two alternatives were proposed:
	- moving the endpoint to `/wp/v2/menu-items/custom`.
	- directly embedding the data, so there wouldn't be the need to create an endpoint.
- Extend or expand the URL dialog so endpoints added by extensions are presented as options in the Page Link block variation. That means there is no need to create new block variations for extension endpoints.
- Use patterns instead of block variations.
- Use a generic Link block that would also allow us to merge other Navigation Link block variations: Page Link, Post Link, etc. So all links, independently of their type, would be added through the same block.

I will need some guidance on moving this forward because I don't think I'm entitled to make a decision on what approach is better. Also, I think it would be great to distinguish what should be fixed as part of this PR and what can be done as follow-ups.

jasmussen commented:
Thanks for all the work here. I'd like to offer thoughts on at least what we should avoid:

- Patterns should always be features we make _in addition_ to a good basic user experience. It's great if a user can use patterns for their entire flow, but it can never be a replacement for a good baseline.
- Extending URL dialogs can likely be an improvement not just to this block but to many others, and likely something we'll want to do regardless of what we do here. Because it's not just in the navigation block where you might want to link to these types of destinations.

At the same time, consolidating on a single Menu Item block doesn't really work that well unless paired with an upgraded URL picker. So to an extent there's a bit of a meta conversation, essentially repeating what @draganescu said:

- Do we think of destination types (post, page, tag, category, cpt, shopping cart) as something that should be available wherever you can make a link?
- Or do we think of each of these as block types?

If we think the latter, then it should be blocks using the existing transform API, so we avoid building stuff just for this one block. Even if we find it necessary to go with blocks (whether for API or pragmatic/move forward reasons), improvements to the quick inserter can help. For example we might want to tailor the default block suggestions, and only surface others when searched for.

@mtias had thoughts on this with the Home Link block, curious what your take is.

Thelmachido commented:
@Aljullu there hasn't been much activity on this PR in a while, just checking if there is still a plan to continue working on this. 

I have removed the Navigation Screen label since the project was moved to [an inactive status on the feature projects page](https://href.li/?https://make.wordpress.org/core/features/) in coordination with the project leads. However, the work can be continued with the navigation block.

Aljullu commented:
> @Aljullu there hasn't been much activity on this PR in a while, just checking if there is still a plan to continue working on this.

I think we can close this PR given the feedback that was provided, I experimented with an alternative approach in #35876. Thanks for the heads-up, @Thelmachido!
